doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,894 @@
+
+## Intro to typing in Python
+
+Python is a dynamically typed (or sometimes called "duck typed") language. This means that the interpreter only checks
+types as code runs and types are allowed to change. This is different from more statically typed languages, like Java,
+where you will have you declare types in code and check them at compile time. However,
+with [PEP 484](https://peps.python.org/pep-0484/), type hints were introduced to Python which makes it possible to do
+the static type checking of Python code. Type hints can be written into Python code to indicate the type of a variable,
+argument, return type, etc. There are tools available to perform static type checking using type hints,
+like [mypy](https://mypy.readthedocs.io/en/stable/). Note that type hints have no effect on the code at runtime and are
+not enforced by the interpreter.
+
+There are some key benefits to using and checking type hints in Python code:
+
+1) type hints checked by a static type checker can help developers find bugs in Python code (type hints can be thought
+   of as "free" unit tests, although they do not replace a thorough test suite).
+2) type hints are shipped in our packages so that customers can consume them to further type check their own
+   applications and get improved IDE and Intellisense experiences
+
+There are two styles of type hints: comment and annotation (preferred).
+
+Type comments are Python 2 compatible and consist of comments in the code which indicate the types of arguments and
+return values. Note that these are truly just comments in the code which the type checker can use.
+
+```python
+from typing import Any, Optional
+
+
+def download_blob_from_url(
+        blob_url,  # type: str
+        output,  # type: str
+        credential=None,  # type: Optional[Any]
+        **kwargs  # type: Any
+):
+    # type: (...) -> None
+    ...
+```
+
+Since Python 3, type annotations were introduced and are the **preferred** method of adding type hints to your code.
+Annotations provide a cleaner syntax and let you keep the type information closer, or inline with the code.
+
+```python
+from typing import Any, Optional
+
+
+def download_blob_from_url(
+        blob_url: str,
+        output: str,
+        credential: Optional[Any] = None,
+        **kwargs: Any
+) -> None:
+    ...
+```
+
+A fully annotated signature includes type annotations for all parameters and the return type. The type of a parameter
+should follow the `:` syntax and a default argument can be supplied like in the `credential` parameter above. A return
+type follows the function def with an arrow `->`, its type, and then `:`.
+
+Note that since Python 3.6, it is also possible to add type annotations to variables. The syntax follows the same as
+function arguments:
+
+```python
+length: int = 42
+width: int = 8
+area: int = length * width
+```
+
+When typing your library, note that almost anything can be used as a type - built-ins like `str`, `int`, `bool`, types
```

Comment:
I remember `list` and `dist` were not usable at first, I don't remember on which Python version they were added, and if `from __future__ import annotations` make them work on 3.6

Reply:
Ok, you replied later, it's from 3.7, so we can't recommend it _yet_